### PR TITLE
Closes #2887 -- adds a typo checker to i for convenience

### LINKED ISSUE #2887: Friendly column finder for "not found" errors
Consider:

```
DT = data.table(myname = 1, some_var = 2)
DT[somevar == 2]
#  Error in eval(.massagei(isub), x, parent.frame()) : 
#   object 'somevar' not found 
```

Should be easy enough to run `adist` on `names(DT)` (or also `ls()`) to suggest some potential misspellings. Nothing too sophisticated of course, maybe just look within levensthein distance of 2 or 3. Truncated to 10 options or so in case there are numbered columns.

### PR DESCRIPTION
**Pull request by MichaelChirico**
Pretty rudimentary, just to get things started.

A few points for discussion/improvement:

 - I eschewed lassoing in objects from `.GlobalEnv` for now, but I left `.checkTypos` flexible enough that it should be easy to expand this (by either sending `c(names(x), ls(envir = .GlobalEnv))` as the `ref` argument, or by splitting into `ref_x` and `ref_global_env` to allow the error to distinguish between objects found in the scope of `x` vs. parent scope)
 - This is only one place in `i`. If we think it's worth it, could add similar code in other places like [here](https://github.com/Rdatatable/data.table/blob/master/R/data.table.R#L394) or even [here](https://github.com/Rdatatable/data.table/blob/master/R/data.table.R#L2017)
 - Nothing done in `j` or `by` for now. 
 - Anyone who's more familiar/accustomed to using `agrep` might be able to suggest different defaults for the `cost` value(s).

### CODE REVIEW

File: man/chmatch.Rd [hunk 1]
```diff
@@ -41,7 +41,8 @@ chgroup(x)
 
 # N is set small here (1e5) because CRAN runs all examples and tests every night, to catch
 # any problems early as R itself changes and other packages run.
-# The comments here apply when N has been changed to 1e8 and were run on 2018-05-13 with R 3.5.0 and data.table 1.11.2.
+# The comments here apply when N has been changed to 1e8
```
Comment (MichaelChirico) [inline man/chmatch.Rd]: this is irrelevant to this PR, but I was getting a `NOTE` from `R CMD check`

**Comment by MichaelChirico** [comment]
agree, we should keep it as experimental in dev for a while.

File: R/data.table.R [hunk 2]
```diff
@@ -200,6 +200,21 @@ replace_dot_alias <- function(e) {
...
 }
 
+.checkTypos = function(err, ref) {
+  if (grepl('object.*not found', err$message)) {
+    used = gsub(".*object '([^']+)'.*", "\\1", err$message)
+    found = agrep(used, ref, value = TRUE, ignore.case = TRUE, fixed = TRUE)
+    if (length(found)) {
+      stop("Object '", used, "' not found. Perhaps you ",
+           sprintf("intended one of: [%s]",
+                   paste(found, collapse = ', ')))
```
Comment (st-pasha) [inline R/data.table.R]: 2 suggestions:
1. Add a special case for `length(found) == 1` ("one of ..." seems awkward for 1 item).
2. Limit the number of columns suggested to 3 or 4 (in theory your DT can have 1000s of columns, similarly named, and all of them might be found by `agrep`).

File: R/data.table.R [hunk 3]
```diff
@@ -200,6 +200,21 @@ replace_dot_alias <- function(e) {
...
+    } else {
+      stop("Object '", used, "' not found among: ",
+           sprintf("[%s]", paste(ref, collapse = ', ')))
```
Comment (st-pasha) [inline R/data.table.R]: Similarly here `ref` can have an enormous size.

**Comment by MichaelChirico** [comment]
a very reasonable case for this is columns like V1,...,Vn or in a wide
table like income00, income01, ..., income18.

I thought of putting in a limit to prevent error explosion, but R does that
itself

https://stackoverflow.com/a/50387968/3576984

OTOH I guess the user should get the point... if they put in Vx and are
matched to V1,...,V99, seeing "V1,V2,V3,..." should be sufficient.

**Comment by mattdowle** [comment]
@MichaelChirico  Did you benchmark to see if this slowed things down?  It's an (expensive) `tryCatch` on most `[.data.table` calls that contains `i` iiuc?  Is there a way to avoid `tryCatch`?

**Comment by MichaelChirico** [comment]
let me investigate...

**Pull request merged**